cranelift/codegen/src/opts/algebraic.isle

```diff
@@ -287,3 +287,11 @@
 (rule (simplify
        (icmp (ty_int ty) (IntCC.SignedLessThanOrEqual) x x))
       (iconst ty (imm64 1)))
+
+;; (x ^ -1) and (-1 ^ x) can be replaced with the `bnot` instruction
+(rule (simplify (bxor ty (iconst ty (u64_from_imm64 0xffff_ffff_ffff_ffff)) x)) (subsume (bnot ty x)))
+(rule (simplify (bxor ty x (iconst ty (u64_from_imm64 0xffff_ffff_ffff_ffff)))) (subsume (bnot ty x)))
```

Comment:
These rules are unlikely to match for any type smaller than I64 because we've been... somewhat... consistent about zeroing the unused bits in `iconst` instructions. If we fix #3059 then these rules would never match for anything but I64.

With the recently-merged #5706, you can write these this way instead:

```suggestion
(rule (simplify (bxor ty (iconst ty k) x))
 (if-let -1 (i64_sextend_imm64 ty k))
 (subsume (bnot ty x)))
(rule (simplify (bxor ty x (iconst ty k)))
 (if-let -1 (i64_sextend_imm64 ty k))
 (subsume (bnot ty x)))
```

The other thing to note though is I think these rules shouldn't use `subsume`. If both operands to `bxor` are constants, then we'd prefer to apply the constant folding rules instead of emitting a `bnot` instruction. If multiple rules use `subsume`, we don't specify which rule will win; it currently depends on details of how ISLE does codegen that are difficult to predict.

But I don't know yet what advice to give about when to use `subsume`. Struggling with that question led me to file #5704 today.

Reply:
To be clear, these rules do match for smaller types in the filetests in this PR because the CLIF text parser is one of the places where we have not been consistent about zeroing the high bits.